### PR TITLE
fuzzgen: Fuzz Switch API

### PR DESCRIPTION
**Pull request by afonso360**
Hey,

Turns out this is an interface that the frontend provides.
We should fuzz it.

cc: @jameysharp

### CODE REVIEW

File: cranelift/fuzzgen/src/function_generator.rs
Comment (bjorn3) [inline cranelift/fuzzgen/src/function_generator.rs:461]: For i128, maybe you could generate two variables and then use `iconcat` if `get_variable_of_type` doesn't support it directly?

**Comment by bjorn3** [comment]
As author of this api, I'm curious if it will be able to find any issues.

File: cranelift/fuzzgen/src/function_generator.rs
Comment (bjorn3) [inline cranelift/fuzzgen/src/function_generator.rs:474]: `Switch` is designed to work for any kind of pattern. It should work just as well for switches like 1..5, 7..12 as for those like 1, 7, 33, 91. The former will use a branch and then two jump tables, while the later will only use branches. I presume the current code is biased towards the later option, right?

File: cranelift/fuzzgen/src/config.rs
Comment (bjorn3) [inline cranelift/fuzzgen/src/config.rs:45]: I think increasing this a bit will allow for more interesting patterns of generated code.

File: cranelift/fuzzgen/src/function_generator.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/function_generator.rs:461]: I'm doing exactly that [here](https://github.com/afonso360/wasmtime/commit/b5802659e58b6eccb6b469483ce4301c11fe6e8b#diff-226e0f113ddbbcedaa5265109a951e52d30aa35307537684f56a5e6bedc9ba0eR323-R328)!

But we need a few additional bits to introduce I128 support.

Right now the fuzzer is hitting #4498 and #4406 when I add I128 support. I can add support for I128 values and work around those two issues by disabling those ops for i128.

I think it would be best to do this in a separate PR.

Comment (bjorn3) [reply cranelift/fuzzgen/src/function_generator.rs:461]: You shouldn't hit https://github.com/bytecodealliance/wasmtime/issues/4498 for the Switch api, right? Only when using `br_table` directly.

File: cranelift/fuzzgen/src/function_generator.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/function_generator.rs:474]: Yeah, Its going to be very unlikely that the fuzzer will generate that. I'll add a bias for adding contiguous ranges.

File: cranelift/fuzzgen/src/function_generator.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/function_generator.rs:461]: I don't think so, but we will hit the special cased lowering. I'll rebase the i128 patch after merging this, and disable the crashing ops so that we can start fuzzing Switch's i128.

**Comment by afonso360** [comment]
We got a crash!

```
test interpret

function %a(i32, i16, i32) -> i8 system_v {
    jt0 = jump_table [block1, block1, block1, block1, block1, block1, block1, block1]
    jt1 = jump_table []

block0(v0: i32, v1: i16, v2: i32):
    v3 = iconst.i32 256
    v4 = iconst.i8 1
    v24 -> v4
    v5 = bconst.b1 false
    v6 = bconst.b1 false
    v7 = bconst.b1 false
    v8 = bconst.b1 false
    v9 = bconst.b1 false
    v10 = bconst.b1 false
    v11 = bconst.b1 false
    v12 = bconst.b1 false
    v13 = bconst.b1 false
    v14 = bconst.b1 false
    v15 = bconst.b1 false
    v16 = bconst.b1 false
    v17 = bconst.b1 false
    v18 = iconst.i64 0
    v19 = iconst.i32 0
    v20 = iconst.i16 0
    v21 = iconst.i8 0
    v22 = uextend.i32 v4
    v23 = icmp_imm eq v22, 0x4100_0000_00bf_d470
    brnz v23, block1
    jump block1

block1:
    br_icmp.i8 eq v24, v24, block1
    jump block1
}
; run: %a(0, 0, 167773184) == 1
```

Fails with:
```
unexpected trap: StepError(ValueError(InvalidInteger(TryFromIntError(()))))
```

This is probably because the imm field is too large for an `i32`

**Review by jameysharp — APPROVED** [review]
This PR has been great, finding bugs in obscure corners before it's even merged. Well done!

With #4510 merged the Switch API should now work for all widths, right? Are there any remaining bugs you know about? I figure I'll wait until Monday to merge this and see if oss-fuzz starts yelling.

**Comment by afonso360** [comment]
> With https://github.com/bytecodealliance/wasmtime/pull/4510 merged the Switch API should now work for all widths, right?

Yeah, We shouldn't need any changes here.

Ill rebase this and leave it fuzzing over the weekend to see if anything else comes up

**Comment by afonso360** [comment]
Fuzzer found nothing new over the weekend with these new changes. Should be ready to merge! Lets see if oss-fuzz finds something else.

**Pull request merged**
